src/stream.rs

```diff
@@ -985,7 +984,33 @@ impl SendBuf {
             return Ok(());
         }
 
-        self.data.push(buf);
+        match self.data.back() {
+            None => self.data.push_back(buf),
+
+            Some(back) =>
+                if buf.off >= back.max_off() {
+                    // Fast path: the new data can simply be appended at the
+                    // end of the send buffer.
+                    self.data.push_back(buf);
+                } else {
+                    // Slow path: the new data needs to be inserted at a
+                    // specific index in the send buffer.
+                    let mut insert_at = None;
+
+                    for i in 0..self.data.len() {
+                        if buf.off < self.data[i].off {
+                            insert_at = Some(i);
+                            break;
+                        }
+                    }
```

Comment:
Maybe a chance to improve this further once binary_search lands? https://github.com/rust-lang/rust/issues/78021

Reply:
The implementation looks pretty simple, so we can probably do it without waiting for that PR. Though note that after the next set of changes (once I get them to work) we won't need the slow path anymore as we will always appending to the end, so might end up being wasted work.